[QE] Improve CRC E2E tests pipelines by adding java application deploy workflow using Eclipse JKube

### Description

As we discussed internally during F2F meeting, CRC can use the [JKube](https://github.com/eclipse-jkube/jkube) project to test the deployment workflow of a simple Java application. 

Once CRC has been set up in E2E tests, we can add one more step to build and deploy a Java application into the installed cluster and verify whether it's accessible from the CRC cluster. [JKube](https://github.com/eclipse-jkube/jkube) performs an S2I binary build to build a container image and then deploys it to the OpenShift cluster.

I had discussed this with @adrianriobo , we can try adding this step in a container.

_originally posted by @adrianriobo  in internal chat:_
> well the basic deploy could be similar to this but thinking on how we could do this...as I said using a container which have access to the cluster will cover an extra use case (maybe a bit edge) but it will cover that type of connectivity..from running container to the cluster

### Expected Behavior
CRC pipelines are improved to verify application deployment workflow for a simple Java application. We can use some simple Quarkus application and deploy it to the created CRC cluster using Eclipse JKube to achieve this.

### Acceptance Criteria
- [ ] basic applications reports 'heatlhy'
- [ ] deploys within 2 mins
- [ ] E2E test written in golang like rest of the integration tests
  - [ ] Test clones Java project (either Spring Boot or Quarkus) 
  - [ ] Test runs maven commands to deploy application
  - [ ] Test performs assertions after application is deployed on these conditions
    - [ ] BuildConfig is created
    - [ ] Build is complete
    - [ ] ImageStream is created
    - [ ] ImageStreamTag is created
    - [ ] Application Pod is ready
    - [ ] Route and Service are created
    - [ ] When application endpoint accessed using go HTTP client, it's accessible and Application is responsive

test on windows and mac where successful

```
> podman run --network host redhat-ubi-openjdk17-with-oc:latest
WARNING: Using insecure TLS client config. Setting this option is not supported!

Login successful.
```

but due to some limitations on windows/mac test do not end:

https://github.com/crc-org/crc/wiki/Podman-support#limitations

> Ports are not automatically exposed on the host.
> 
> Workaround when using vsock network mode:
> Expose a port: curl  --unix-socket ~/.crc/crc-http.sock http:/unix/network/services/forwarder/expose -X POST -d '{"local":":8080","remote":"192.168.127.3:8080"}'

Results from running:

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.eclipse.jkube.integrationtests.quarkus.rest.QuarkusOcITCase
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[pool-3-thread-2] WARN org.eclipse.jkube.integrationtests.assertions.KubernetesClientAssertion - Connection to http://localhost:30633/ failed, retrying
[ForkJoinPool-1-worker-1] INFO org.eclipse.jetty.websocket.client.WebSocketClient - Shutdown WebSocketClient@63f1c585[coreClient=WebSocketCoreClient@4886affc{STARTED},openSessions.size=0]
[ERROR] Tests run: 5, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 186.8 s <<< FAILURE! -- in org.eclipse.jkube.integrationtests.quarkus.rest.QuarkusOcITCase
[ERROR] org.eclipse.jkube.integrationtests.quarkus.rest.QuarkusOcITCase.ocApply -- Time elapsed: 102.3 s <<< ERROR!
java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1960)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2095)
	at org.eclipse.jkube.integrationtests.assertions.ServiceAssertion.assertNodePortResponse(ServiceAssertion.java:109)
	at org.eclipse.jkube.integrationtests.quarkus.rest.Quarkus.assertThatShouldApplyResources(Quarkus.java:68)
	at org.eclipse.jkube.integrationtests.quarkus.rest.QuarkusOcITCase.ocApply(QuarkusOcITCase.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   QuarkusOcITCase.ocApply:103->Quarkus.assertThatShouldApplyResources:68 » Timeout
[INFO] 
[ERROR] Tests run: 5, Failures: 0, Errors: 1, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-failsafe-plugin:3.1.2:verify (default) @ integration-tests ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for JKube :: Integration Tests :: Parent 0.0.0-SNAPSHOT:
[INFO] 
[INFO] JKube :: Integration Tests :: Parent ............... SUCCESS [  0.002 s]
[INFO] JKube :: Integration Tests :: Quarkus :: Rest ...... SUCCESS [ 16.386 s]
[INFO] JKube :: Integration Tests :: Quarkus :: Rest :: Trace Logging Enabled SUCCESS [  3.390 s]
[INFO] JKube :: Integration Tests :: Tests ................ FAILURE [03:08 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:37 min
[INFO] Finished at: 2024-10-15T10:53:37Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:3.1.2:verify (default) on project integration-tests: 
[ERROR] 
[ERROR] Please refer to /home/default/jkube-integration-tests/it/target/failsafe-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :integration-tests
```

@rohanKanojia Is it possible to fix that port and share?

--- a/test/e2e/features/application_deployment.feature
+++ b/test/e2e/features/application_deployment.feature
@@ -1,0 +1,37 @@
+@story_application_deployment
+Feature: Application Deployment Test
+
+    User deploys a basic java application onto CRC cluster and expects that it's
+    deployed successfully and is accessible via route
+
+    Background:
+        Given ensuring CRC cluster is running
+        And ensuring oc command is available
+        And ensuring user is logged in succeeds
+
+    @testdata @linux @windows @darwin @cleanup @needs_namespace
+    Scenario: Deploy a java application using Eclipse JKube in pod and then verify it's health
+        When executing "oc new-project testproj" succeeds
+        And executing "oc create -f jkube-kubernetes-build-resources.yaml" succeeds
+        And executing "oc start-build jkube-application-deploy-buildconfig --follow" succeeds
+        And executing "oc rollout status -w dc/jkube-application-deploy-test --timeout=600s" succeeds
+        Then stdout should contain "successfully rolled out"
+        And executing "oc logs -lapp=jkube-application-deploy-test -f" succeeds
+        And executing "oc rollout status -w dc/quarkus --timeout=600s" succeeds
+        Then stdout should contain "successfully rolled out"
+        And executing "oc get build -lapp=quarkus" succeeds
+        Then stdout should contain "quarkus-s2i"
+        And executing "oc get buildconfig -lapp=quarkus" succeeds
+        Then stdout should contain "quarkus-s2i"
+        And executing "oc get imagestream -lapp=quarkus" succeeds
+        Then stdout should contain "quarkus"
+        And executing "oc get pods -lapp=quarkus" succeeds
+        Then stdout should contain "quarkus"
+        And stdout should contain "1/1     Running"
+        And executing "oc get svc -lapp=quarkus" succeeds
+        Then stdout should contain "quarkus"
+        And executing "oc get routes -lapp=quarkus" succeeds
+        Then stdout should contain "quarkus"
+        And with up to "4" retries with wait period of "1m" http response from "http://quarkus-testproj.apps-crc.testing" has status code "200"
+        Then executing "curl -s http://quarkus-testproj.apps-crc.testing" succeeds
+        And stdout should contain "{"applicationName":"JKube","message":"Subatomic JKube really whips the llama's ass!"}"

--- a/test/testdata/jkube-kubernetes-build-resources.yaml
+++ b/test/testdata/jkube-kubernetes-build-resources.yaml
@@ -1,0 +1,146 @@
+apiVersion: v1
+kind: List
+items:
+  - apiVersion: rbac.authorization.k8s.io/v1
+    kind: Role
+    metadata:
+      name: jkube-application-deploy-role
+      labels:
+        app: jkube-application-deploy-test
+    rules:
+      - apiGroups:
+          - ""
+        resources:
+          - secrets
+          - pods
+          - pods/log
+          - services
+          - events
+        verbs:
+          - create
+          - get
+          - list
+          - update
+          - watch
+          - patch
+      - apiGroups:
+          - apps.openshift.io
+        resources:
+          - deploymentconfigs
+        verbs:
+          - create
+          - get
+          - list
+          - update
+          - patch
+      - apiGroups:
+          - build.openshift.io
+        resources:
+          - buildconfigs
+          - buildconfigs/instantiatebinary
+          - builds
+        verbs: ["*"]
+      - apiGroups:
+          - image.openshift.io
+        resources:
+          - imagestreams
+        verbs:
+          - create
+          - get
+          - list
+          - update
+          - patch
+      - apiGroups:
+          - route.openshift.io
+        resources:
+          - routes
+        verbs:
+          - create
+          - get
+          - list
+          - update
+          - patch
+  - apiVersion: rbac.authorization.k8s.io/v1
+    kind: RoleBinding
+    metadata:
+      name: jkube-application-deploy-binding
+      labels:
+        app: jkube-application-deploy-test
+    roleRef:
+      apiGroup: rbac.authorization.k8s.io
+      kind: Role
+      name: jkube-application-deploy-role
+    subjects:
+      - kind: ServiceAccount
+        name: jkube-application-deploy-sa
+  - apiVersion: v1
+    kind: ServiceAccount
+    metadata:
+      name: jkube-application-deploy-sa
+      labels:
+        app: jkube-application-deploy-test
+  - apiVersion: image.openshift.io/v1
+    kind: ImageStream
+    metadata:
+      labels:
+        app: jkube-application-deploy-test
+      name: jkube-application-deploy-is
+  - apiVersion: build.openshift.io/v1
+    kind: BuildConfig
+    metadata:
+      name: jkube-application-deploy-buildconfig
+      labels:
+        app: jkube-application-deploy-test
+    spec:
+      output:
+        to:
+          kind: ImageStreamTag
+          name: jkube-application-deploy-is:latest
+      source:
+        type: Dockerfile
+        dockerfile: |
+          FROM registry.access.redhat.com/ubi9/openjdk-17:1.20-2.1726695177
+          LABEL org.opencontainers.image.authors="CRCQE <devtools-crc-qe@redhat.com>"
+          USER root
+          # Install Git
+          RUN microdnf install -y git
+          RUN git clone https://github.com/eclipse-jkube/jkube.git
+          RUN chmod -R 775 /home/default/jkube
+          WORKDIR /home/default/jkube/quickstarts/maven/quarkus
+          RUN mkdir foo
+          ENTRYPOINT ["mvn", "package", "oc:build", "oc:resource", "oc:apply"]
+      strategy:
+        type: Docker
+  - apiVersion: apps.openshift.io/v1
+    kind: DeploymentConfig
+    metadata:
+      labels:
+        app: jkube-application-deploy-test
+      name: jkube-application-deploy-test
+    spec:
+      replicas: 1
+      selector:
+        app: jkube-application-deploy-test
+      template:
+        metadata:
+          labels:
+            app: jkube-application-deploy-test
+          name: jkube-application-deploy-test
+        spec:
+          containers:
+            - image: jkube-application-deploy-is:latest
+              imagePullPolicy: IfNotPresent
+              name: maven-pod
+              securityContext:
+                privileged: false
+          serviceAccount: jkube-application-deploy-sa
+      triggers:
+        - type: ConfigChange
+        - imageChangeParams:
+            automatic: true
+            containerNames:
+              - maven-pod
+            from:
+              kind: ImageStreamTag
+              name: jkube-application-deploy-is:latest
+          type: ImageChange

